### PR TITLE
Remove legacy join APIs

### LINKED ISSUE #7762: [BUG] Legacy join APIs return extraneous columns
**Describe the bug**
While updating the Java bindings to the new join gather map APIs added in #7454, I noticed that the old join APIs now return more columns than they did before.  They seem to be returning a table that contains the key columns from both the left and right tables, whereas before they returned only the key columns as they appeared in the left table (updated for the join result).

For example take two tables, the left having columns A,B and the right having columns C,D and we want to perform a join on A == C.  Previously the join APIs would return a table with 3 columns, A, B, and D, but now they return 4 columns.

**Steps/Code to reproduce bug**
Call the `left_join`, `inner_join`, and `full_join` APIs before and after #7454.

**Expected behavior**
Redundant key columns should not be manifested.  Manifesting the right key columns on an inner or left join seems to serve no purpose, as the left key columns already contain the proper answer for the join.

### PR DESCRIPTION
**Pull request by vyasr**
Resolves #7762

### CODE REVIEW

**Comment by vyasr** [comment]
@jlowe I've removed the non-gather map JNI functions for computing left semi/anti joins since there are gather map versions and I assumed that everything would have been switched to use those, but if that's not the case please let me know what you'd like to do here.

**Comment by jlowe** [comment]
The JNI code was removed, but the corresponding cudf Java code referring to that JNI code (i.e.: the `Table.leftSemiJoin` and `Table.leftAntiJoin` methods) have not been removed as well.  There's also a number of unit tests in TableTest that will need to be removed as well.

Also note that there is Java code for non-gather map versions of left/right/inner/full joins which arguably should be removed along with the corresponding JNI/unit tests if we're removing all the non-gather-map APIs (e.g.: `Table.leftJoin`, `Table.rightJoin`, `Table.innerJoin`, and `Table.fullJoin`).  Unlike the semi join forms of the non-gather map APIs, these APIs use the gather-map forms in the JNI code because a while back the non-gather-map APIs were arguably broken when they returned the left and right join keys for _each_ table, redundantly.  That's what led me to file #7762.  If you'd rather have that done as a separate followup, let me know.

I verified the RAPIDS Accelerator for Apache Spark does not use any non-gather-map APIs, so removing the non-gather-map join APIs should not pose a problem for that application.

**Comment by ttnghia** [comment]
So now we only have the public join APIs returning the gather map. I wonder why these gather maps have to be wrapped in `unique_ptr<>`?

**Comment by jlowe** [comment]
> I wonder why these gather maps have to be wrapped in unique_ptr<>?

The gather maps are being GPU-allocated by the function and thus need to be tracked as a resource that will be freed on destruction.  Returning a `std::unique_ptr` is how that's typically done, similar to how `cudf::table` and `cudf::column` are handled in other libcudf APIs.

File: java/src/main/native/src/TableJni.cpp [hunk 1]
```diff
@@ -2025,148 +2025,6 @@ JNIEXPORT void JNICALL Java_ai_rapids_cudf_Table_readArrowIPCEnd(JNIEnv *env, jc
...
-    auto num_join_rows = static_cast<cudf::size_type>(join_maps.first->size());
-
-    // compute the join result rows for the left table columns
-    auto left_gather_col = cudf::column_view(cudf::data_type{cudf::type_id::INT32}, num_join_rows,
-                                             join_maps.first->data());
-    auto left_out_table =
-        cudf::gather(*left_in_table, left_gather_col, cudf::out_of_bounds_policy::DONT_CHECK);
-
-    // compute the join result rows for the right table columns
-    auto right_non_keys_table = cudf::jni::get_non_keys_table(right_in_table, right_key_indices);
```
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:2060]: `get_non_keys_table` defined earlier in this file is now no longer used.

File: java/src/main/native/src/TableJni.cpp [hunk 2]
```diff
@@ -2025,148 +2025,6 @@ JNIEXPORT void JNICALL Java_ai_rapids_cudf_Table_readArrowIPCEnd(JNIEnv *env, jc
...
-  JNI_NULL_CHECK(env, j_left_key_indices, "left_col_join_indices is null", NULL);
-  JNI_NULL_CHECK(env, j_right_table, "right_table is null", NULL);
-  JNI_NULL_CHECK(env, j_right_key_indices, "right_col_join_indices is null", NULL);
-
-  try {
-    cudf::jni::auto_set_device(env);
-    auto left_in_table = reinterpret_cast<cudf::table_view *>(j_left_table);
-    auto right_in_table = reinterpret_cast<cudf::table_view *>(j_right_table);
-    cudf::jni::native_jintArray left_key_indices(env, j_left_key_indices);
-    auto left_keys_table = cudf::jni::get_keys_table(left_in_table, left_key_indices);
```
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:2041]: `get_keys_table` defined earlier in this file is now no longer used.

File: java/src/main/native/src/TableJni.cpp [hunk 3]
```diff
@@ -2025,148 +2025,6 @@ JNIEXPORT void JNICALL Java_ai_rapids_cudf_Table_readArrowIPCEnd(JNIEnv *env, jc
...
-
-    // compute the join result rows for the right table columns
-    auto right_non_keys_table = cudf::jni::get_non_keys_table(right_in_table, right_key_indices);
-    right_key_indices.cancel();
-    auto right_gather_col = cudf::column_view(cudf::data_type{cudf::type_id::INT32}, num_join_rows,
-                                              join_maps.second->data());
-    auto right_out_table =
-        cudf::gather(right_non_keys_table, right_gather_col, cudf::out_of_bounds_policy::NULLIFY);
-
-    return cudf::jni::combine_join_results(env, *left_out_table, *right_out_table);
```
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:2067]: `combine_join_results` defined earlier in this file is now no longer used.

File: cpp/tests/join/semi_anti_join_tests.cpp [hunk 4]
```diff
@@ -40,6 +40,56 @@ using Table          = cudf::table;
 struct JoinTest : public cudf::test::BaseFixture {
 };
 
+// This function is a wrapper around cudf's join APIs that takes the gather map
+// from join APIs and materializes the table that would be created by gathering
+// from the joined tables. Join APIs originally returned tables like this, but
+// they were modified in https://github.com/rapidsai/cudf/pull/7454. This
+// helper function allows us to avoid rewriting all our tests in terms of
+// gather maps.
+template <std::unique_ptr<rmm::device_uvector<cudf::size_type>> (*join_impl)(
```
Comment (ttnghia) [inline cpp/tests/join/semi_anti_join_tests.cpp:49]: Please wrap your utility functions in source files in an anonymous namespace.

**Review by ttnghia — APPROVED** [review]
C++ approval with a recommended change.

**Review by harrism — APPROVED** [review]

**Review by hyperbolic2346 — APPROVED** [review]
So much red. This feels so good to see.

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**